## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [49]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [50]:
# View the column names in the data
column_names = df.columns
print(column_names)
#Or
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [51]:
# Use the describe function to gather some basic statistics
basic_stats = df.describe()
print(basic_stats)


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [53]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# 1. Take a look at my data getting the first and last 5 row in the dataframe 
df.head()
df.tail()

# 2. Display the column names
df.columns

# 3. Display the datatypes 
data_types = df.dtypes
print(data_types)

# 4. Identify missing values. 
missing_values = df.isnull().sum()
print(missing_values)

# 5. Display the statistic of the data
print(basic_stats)

# 6. Display the unique values of a column
unique_values = df['category'].value_counts()
print(unique_values)

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object
first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64
          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438

In [54]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().nlargest
print(top_categories)


<bound method Series.nlargest of category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64>


In [56]:
# For the category with the most entries,
top_categories = df['category'].value_counts().nlargest(1)
print(top_categories)

# which subcategory had the most entries?
top_subcategories = df['subcategory'].value_counts().nlargest(1)
print(top_subcategories)

#For the category with the most entries, which subcategory had the most entries?
top_categories = df['category'].value_counts().idxmax()
top_categories_df = df[df['category'] == top_categories]
top_subcategories = top_categories_df['subcategory'].value_counts().nlargest(1)

print("Top Category:", top_categories)
print("Top Subcategories:",top_subcategories )



category
consumables    23538
Name: count, dtype: int64
subcategory
bathroom supplies    6424
Name: count, dtype: int64
Top Category: consumables
Top Subcategories: subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [62]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().nlargest(5)
print("Top 5 clients with the most entries:", top_clients)


Top 5 clients with the most entries: client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [63]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().nlargest(5)
top_5_client_ids = top_5_clients.index.tolist()
print("Top five clients IDs with the most entries:", top_5_client_ids)

Top five clients IDs with the most entries: [33615, 66037, 46820, 38378, 24741]


In [64]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()
top_client_df = df[df['client_id'] == top_client]
total_qty_top_client = top_client_df['qty'].sum()

print("Top client with the most entries:", top_client)
print("Total quantity ordered by the top client:", total_qty_top_client)



Top client with the most entries: 33615
Total quantity ordered by the top client: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [67]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['line_subtotal'] = df['unit_price'] * df['qty'] 
desired_columns = ['line_subtotal','unit_price','qty']
subset_df = df[desired_columns]
print(subset_df)

       line_subtotal  unit_price   qty
0          115164.00     1096.80   105
1             523.95       24.95    21
2             527.28       13.52    39
3            1056.18       36.42    29
4            3902.00      195.10    20
...              ...         ...   ...
54634        2743.29       83.13    33
54635        9709.73      206.59    47
54636       31188.50       65.66   475
54637         165.76        1.48   112
54638        3103.31        3.01  1031

[54639 rows x 3 columns]


In [70]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calculate_shipping_price(row):
    if row['qty'] > 50:
        return 7 * row['qty'] # $7 per lb for orders over 50 lbs.
    else:
        return 10 * row['qty'] # $10 per lb for orders under 50 lbs. 
    
df['shipping_price'] = df.apply(calculate_shipping_price, axis=1)

print(df[['qty', 'shipping_price']])


        qty  shipping_price
0       105             735
1        21             210
2        39             390
3        29             290
4        20             200
...     ...             ...
54634    33             330
54635    47             470
54636   475            3325
54637   112             784
54638  1031            7217

[54639 rows x 2 columns]


In [81]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
def calculate_total_price(row):
    line_subtotal = row['line_subtotal'] 
    shipping_price = row['shipping_price']
    sales_tax = (line_subtotal + shipping_price) * 0.0925 # Calculate sales tax (9.25%)
    total_price = line_subtotal + shipping_price + sales_tax 
    return total_price
df['total_price'] = df.apply(calculate_total_price, axis=1)

print(df[['line_subtotal', 'shipping_price','total_price': .format(total_price)']])



SyntaxError: unterminated string literal (detected at line 12) (1206768774.py, line 12)

In [72]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']])



       unit_cost   qty  shipping_price  line_cost
0         762.71   105             735   80819.55
1          15.09    21             210     526.89
2           7.86    39             390     696.54
3          24.85    29             290    1010.65
4         108.17    20             200    2363.40
...          ...   ...             ...        ...
54634      51.60    33             330    2032.80
54635     175.46    47             470    8716.62
54636      57.31   475            3325   30547.25
54637       1.22   112             784     920.64
54638       1.61  1031            7217    8876.91

[54639 rows x 4 columns]


In [77]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['shipping_price'] - df['line_cost']
print(df)

           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0       

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [80]:
# Check your work using the totals above

# Filter the DataFrame for Order ID 2742071
order_2742071 = df[df['order_id'] == 2742071]

# Check if the total price for Order ID 2742071 matches the expected value
expected_total_price = 152811.89
confirmed = order_2742071['total_price'].sum() == expected_total_price

if confirmed:
    print("Order ID 2742071 had a total price of $152,811.89")
else:
    print("Total price for Order ID 2742071 does not match the expected value")


# Filter the DataFrame for Order ID 2173913
order_2173913 = df[df['order_id'] == 2173913]

# Check if the total price for Order ID 2742071 matches the expected value
expected_total_price = 162388.71
confirmed = order_2173913['total_price'].sum() == expected_total_price

if confirmed:
    print("Order ID 2173913 had a total price of $162388.71")
else:
    print("Total price for Order ID 2173913 does not match the expected value")


# Filter the DataFrame for Order ID 6128929
order_6128929 = df[df['order_id'] == 6128929]

# Check if the total price for Order ID 6128929 matches the expected value
expected_total_price = 923441.25
confirmed = order_6128929['total_price'].sum() == expected_total_price

if confirmed:
    print("Order ID 6128929 had a total price of $923441.25")
else:
    print("Total price for Order ID 6128929 does not match the expected value")



Total price for Order ID 2742071 does not match the expected value
Total price for Order ID 2173913 does not match the expected value
Total price for Order ID 6128929 does not match the expected value


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M